### DISCUSSION TITLE
Cardinal Host - Non LV2 support

### DISCUSSION THREAD

**Discussion by jp-gorman** [DISTRHO/Cardinal#105] [Q&A]
I noted support for LV2 modules and Cardinal internal plugins in the Cardinal Host modeule. Do you envisage support for other VST, AU plugins (simlar to what the current Rack Host offers?) I tried Cardinal within Cardinal and worked out how to get inner audio out of it, not that this would be needed as side by side makes more sense, but it was a test for any plugins, just wondered what was left to get other plugins working natively within cardinal.

**Comment by falkTX** [comment]
You can use Carla which supports a bunch of plugin formats, it is available as a module inside Cardinal, double-click on it to show its GUI.
It requires Carla (the application) to be installed system-wide though, something that is not possible on Windows for now (but works on macOS and other systems)

Getting VST and others to work within Ildaeil (the host module you are using) is possible, in due time.
It is just that, contrary to LV2, VST and others do not have a way to nicely lookup for the existing plugins quickly and safely. VST etc require us to implement binary scanning of plugins, which I didnt implement in Ildaeil.
At some point I want to have some sort of cached DB file that contains information about available plugins, or alternatively read that information from Carla Qt config stuff.

If you know how to code and want to help out, I can guide you on the steps needed for it.

**Reply by jp-gorman** [reply]
Installing Carla now and will use that to check out the VSTs on OSX.
What would the advantages of Ildaeil over Carla - Providing Windows support for VSTs?
I'll dig a bit more into the project before I get involved in any coding and perhaps bush up on my C++ with a VCV module or 2 first ;)
Thanks for the fast input. I forced myself not to get VCVRack Pro as I think a free VST and Host is essetial. No worries people profiting off best in breed modules, but basic hosting and DAW interaction needs to be open source. Great work puling all this together.

**Reply by falkTX** [reply]
Thank you.

For Carla, the case is just to be able to have a full blown proper modular host vs Ildaeil which is a tiny abstraction of it.
Carla can do plugin scanning, drag&drop plugin files, internal patchbay and all sorts of stuff a modular plugin host has.

Ildaeil is there just because using Carla is sometimes overkill, for example when we want to run a single simple plugin.

**Reply by jp-gorman** [reply]
I installed Carla and was able to load it into ut's GUI seperatly from console. I could only find a carla host module in the Cardinal Library. I added it but double clicking does nothing and there is no similar plugin loading and hence GUI launching capabilty like I saw when I loaded the Ildaeil Host and it's plugins. I must be missing something ;(

**Reply by falkTX** [reply]
check the host logs to see if something appears, I assume you are able to launch the carla application standalone, correct?
maybe starting carla manually first is needed, macOS tends to be quite picky in regards to what users are allowed to do.

**Reply by jp-gorman** [reply]
OK, having no clue about Carla here's what I see so far. Loading Carla seperately and adding Cardinal as a plugin, allows me to see cardinal UI on double click of the PatchBay Cardinal. I add in a clock-ed-->and addr seq --> fm-op -->   linked the Audio 1 Input of Carla Plugin Host. In Carla I have The Cardinal Synth Plugin Audio Output 1&2 connected to Audio Output 1&2 (but I hear nothing).
I tested the direct cable used here with audio8 plugin on the Rack instance and heard the CV altering.

So....

Is every instance of Carla host in a Cardinal Rack Module, linked to the global system wide Carla? i.e. if I host Cardinal in Reaper or in Carla itself? Or does carla only see the plugins loaded into itself...if so then is linking VST plugins etc. to Cardinal only available once loaded into a carla instance of the Cardinal Plugin (and hence cardinal in reaper as a plugin is pointless)

Currently in both cases I get no sound in main system when routing from Cardinal (hosted in Carla) to "Audio Output 1&2" from Cardinal Synth "Audio Output 1&2"

**Reply by falkTX** [reply]
Carla inside Cardinal inside Carla is a bit of a weird setup, so yeah that gets confusing.
Cardinal does have its own internal version of Carla, but the barebones to get audio working. What is missing from the Cardinal package is Carla's frontend - so that double-click actually opens something.

The frontend of Carla is actually a separate application, which talks to the backend over unix pipes.
This is how we can separate the 2, and later then still join them and it still works.

When you load Cardinal in reaper, then load Carla inside Cardinal and double-click on it, it will open /Applications/Carla.app contents in order to show the full GUI.



> Currently in both cases I get no sound in main system when routing from Cardinal (hosted in Carla) to "Audio Output 1&2" from Cardinal Synth "Audio Output 1&2"

You might just need to tweak the Carla engine settings to have Carla use your soundcard.

**Comment by jp-gorman** [comment]
Ok To clarify cause I've missed you. In reaper or Carla as the base application UI host, I load Cardinal and put a Carl's host module into it. It's that Carla host in Rack that I assume is the link from Racks CV and audio ports to the Carla UI that can have my other plugins (vst etc) (i.e we cannot have plugins directly in Rack, only linked ports to the actual plugins running in Carla UI.
So, in this Instance no "inception"... I just linked the ports from Carla UI Cardinal module to audio outs in Carla (which produced no sound)
This sounds like Cardinal in reaper is pointless if I need to open a seperate Carla UI (I e. It does not open when Carla host is doable clicked from Cardinal). Am I correct?
I'll try tweak the settings in Carla and read up a bit. This is why I prefer Cardinal over rack Pro...I'm already learning about Carla and LVM2 plugin creation ;)

Sent from Yahoo Mail on Android 
 
  On Sun, 9 Jan 2022 at 2:26, Filipe ***@***.***> wrote:   


Carla inside Cardinal inside Carla is a bit of a weird setup, so yeah that gets confusing.
Cardinal does have its own internal version of Carla, but the barebones to get audio working. What is missing from the Cardinal package is Carla's frontend - so that double-click actually opens something.

The frontend of Carla is actually a separate application, which talks to the backend over unix pipes.
This is how we can separate the 2, and later then still join them and it still works.

When you load Cardinal in reaper, then load Carla inside Cardinal and double-click on it, it will open /Applications/Carla.app contents in order to show the full GUI.


Currently in both cases I get no sound in main system when routing from Cardinal (hosted in Carla) to "Audio Output 1&2" from Cardinal Synth "Audio Output 1&2"


You might just need to tweak the Carla engine settings to have Carla use your soundcard.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Reply by falkTX** [reply]
You can ignore the standalone Carla for now, the double-click on Carla's Cardinal module is supposed to start up the UI.
Why that fails I still dont know.

If you run reaper from the terminal I am quite sure there will be something in there.

I just tried this here with FLStudio on macOS.
Works as intended:

<img width="1280" alt="image" src="https://user-images.githubusercontent.com/1334853/148674380-c5d0ceab-c3ba-42b9-8aad-4bd975b8e458.png">

When we double-click on the cardinal carla module, the full carla gui opens up.
Then we can do some connections to get audio flowing. Both in Carla's internal side, and in Cardinal too.

**Comment by jp-gorman** [comment]
Yes, I can confirm running Carla from the console shows that the host (acording to Reaper logs is:) "Trying to start carla-plugin using "/Applications/Carla.app/Contents/MacOS/resources/carla-plugin-patchbay". It fails to fina and I smy linked to my carla binary at "/usr/local/bin/carla". IT then tries to load cara and throws an error "Failed to load project - Requested file does not exist or is not a readible file" - Choosing the only option "ok" kills the cara that attempted to load. Any ideas on the correct linking/parameters?

**Reply by falkTX** [reply]
I dont know what you are doing, but it is surely wrong.
Carla.app must be installed either in /Applications or in the user specific ~/Applications

There is no other choice.

**Reply by jp-gorman** [reply]
Yup, given that verification, I uninstalled my brew installed "Carla" and found a universal installer at https://kx.studio/Applications:Carla. (for any future users). Dropping these .app files in the Appliations folder made it all work ;)
The cool thing is this version has the CV In's and Out's which the prior verion did not - Happy days.

Thanks for the help.

**Reply by falkTX** [reply]
I was unaware of this homebrew version! I need to install it and do some tests to see if it is possible to reuse
